### ISSUE TITLE
Add madvise with MADV_DONTNEED at GCToOSInterface::VirtualDecommit

### ISSUE THREAD

**Issue by aviviadi** [dotnet/coreclr#15790]
Suggestion:

In order to help the OS even better to understand we don't need the memory anymore, I think madvise(address, size, MADV_DONTNEED) in addtion to the mmap call is needed.

Here : https://github.com/dotnet/corert/blob/f9bf1fc506347c30357a0d708bf69459f142eee4/src/Native/gc/unix/gcenv.unix.cpp#L338

**Comment by ayende** [comment]
To give more details. We got a bug report that our application uses 1.7 GB of working set (out of total of 2GB on the machine) while the actual managed memory was just 300 MB.

This was discovered by looking at the system call and realizing the `RetainVM` flag had a huge impact on the working set we use, even though the amount of managed memory is small.
When not using the retain vm flag, the working set and memory used are much closer.

We tested this and `PROT_NONE` seems to remove the data from the working set properly, so we are puzzled as to why this isn't the case with `RetainVM`. I'm at least assuming that this is what it will end up calling eventually.

**Comment by jkotas** [comment]
cc @swgillespie

**Comment by swgillespie** [comment]
Sounds like there are two separate things here: 1) working set is huge when using `RetainVM` and 2) not using madvise when decommitting segments.

As far as 1) goes, this sounds like it was by design. The GC tries *very* hard to avoid decommitting segments when `RetainVM` is set, going to pretty extreme lengths (particularly with server GC). My understanding is that this was for web services that were thrashing in the memory manager due to repeated commits and decommits of the same heap segments; instead of decommitting a segment that's empty, the GC may opt to "hoard" it and keep it committed for the next time we need a new segment. Each heap does this, so there may be hoarded segments in play for every server GC heap.

For 2), my understanding is that `madvise` with `MADV_DONTNEED` and [what we are doing today](https://github.com/dotnet/coreclr/blob/064136ca259dcf47dc8cb34516a44ba057eb01f1/src/pal/src/map/virtual.cpp#L1512-L1516) are functionally equivalent. Is this not the case? It looks to me that Linux's implementation of `MADV_DONTNEED` will eagerly [unmap the pages in question](https://github.com/torvalds/linux/blob/f6f3732162b5ae3c771b9285a5a32d72b8586920/mm/memory.c#L1596-L1622) and invalidate the TLB entries, which we'll need to do with `mmap` with `PROT_NONE` anyway. Is there a perf advantage to madvising before remapping the region? My intuition is that madvise and mmap together together end up doing the same amount of work that mmap is already doing.

**Comment by ayende** [comment]
@swgillespie Okay, the root of the issue is that as we are starting to hit the memory limit, the segments that the GC is  not using are the ones that are holding up about 60% of the used working set.
Note that this is the working set that really worries me. Because the process is using so much memory, we start to run into low memory protection and everything grinds to a halt.

I'm all for not trashing the memory, but my worry is that in this scenario, we are seeing high memory usage even though we aren't actually using a lot of memory. 

In addition to committing / de-committing the pages, I think that when the GC move them to the pending list, it can call `MADV_FREE`, which keep the allocation, but give the kernel the hint that it isn't used.

Basically, our concern is the usage of too much memory when we are nearing the total of the system usable space, where most of it is in the GC retained buffers.

We currently recommend users will disable this setting on machines with low amount of memory, but it would be nice if we could have better behavior.

**Comment by swgillespie** [comment]
@ayende Isn't RetainVM off by default? I would be surprised to hear that it's on by default, historically it hasn't been.

**Comment by ayende** [comment]
It's off by default, yes.

It's a story with a bit of history. When we run performance testing we noticed that we got a major boost from `RetainVM`. We were testing on a machine with a lot of memory, aiming for common production scenarios.

Then came docker, and now it is common for our users to run us on instances with minimal amount of memory. The impetus for this issue is that we run into "using too much memory" issue and traced it back to this flag.

We _can_ (and have) set this flag to false in this situation, but I would rather have the cake and eat it too if this is possible and have this setting also handle this situation as welll

**Comment by swgillespie** [comment]
cc @Maoni0

**Comment by Maoni0** [comment]
RetainVM should not affect how GC commits memory; it should only affect how GC reserves memory, ie, it will keep the memory reserved for existing segments (if they are not larger than normal) so we don't lose that VA range. but we will still decommit memory. this was added to combat VA fragmentation on 32-bit. so it's interesting you are seeing this affecting your committed charge - it shouldn't. this should be debugged. if you have 2 dumps, one with the RetainVM option and one without and could share with us, that'd be great. would that be possible?

**Comment by ayende** [comment]
Yes, that is something that we can certainly do.
@aviviadi Can you send them?

**Comment by aviviadi** [comment]
Trying to figure out what makes our RSS so high - apparently this is not caused by the "System.GC.RetainVM".. it is because of "System.GC.Server" when set to true.

The following program demonstrate the differences
With runtimeconfig : 
```
"configProperties": {
      "System.GC.Concurrent": true,
      "System.GC.Server": false,           <<<<<< false
      "System.GC.RetainVM": true
    }
```
Output:
```
Working set: 29,315,072
Managed mem: 226,408
Working set: 1,153,503,232
Managed mem: 1,107,659,584
Working set: 43,110,400     <<<=============== After collecting
Managed mem: 162,928
```
And with runtimeconfig:
```
"configProperties": {
      "System.GC.Concurrent": true,
      "System.GC.Server": true,           <<<<< true
      "System.GC.RetainVM": true
    }
```
Output:
```
Working set: 37,265,408
Managed mem: 243,464
Working set: 1,172,680,704
Managed mem: 1,115,420,608
Working set: 774,225,920     <<<=========== After collecting
Managed mem: 172,752
```

So, when the RSS is high, OOM Killer lays eyes on that process.. And this memory is considered as recently used by the OS. 


The program: (Executed on Ubuntu 16.04, dotnet 2.0.5)
```
using System;
using System.Collections.Generic;
using System.ComponentModel.Design.Serialization;
using System.Diagnostics;
using System.Threading.Tasks;

namespace ConsoleApp1Rvm
{
    class Program
    {
        static void Main(string[] args)
        {
            var array = new List<byte[]>[64];
            var p = Process.GetCurrentProcess();
            Console.WriteLine("Working set: " + p.WorkingSet64.ToString("#,#;;0"));
            Console.WriteLine("Managed mem: " + GC.GetTotalMemory(false).ToString("#,#;;0"));
            for (int i = 0; i < 64; i++)
            {
                array[i] = new List<byte[]>();
            }
            var r = new Random(123);
            for (int i = 0; i < 1024; i++)
            {
                for (int j = 0; j < 1024; j++)
                {
                    var bytes = new byte[1024];
                    r.NextBytes(bytes);
                    array[i%array.Length].Add(bytes);
                }

                //array[r.Next(0, args.Length)].Clear();
            }

            Parallel.For(0, 64, i =>
            {
                Parallel.ForEach(array[i], b =>
                {
                    for (int j = 0; j < b.Length; j++)
                    {
                        b[i] ^= 1;
                    }
                });
            });
            p.Refresh();
            Console.WriteLine("Working set: " + p.WorkingSet64.ToString("#,#;;0"));
            Console.WriteLine("Managed mem: " + GC.GetTotalMemory(false).ToString("#,#;;0"));
            array = null;

            GC.Collect(2);
            GC.WaitForPendingFinalizers();
            p.Refresh();
            Console.WriteLine("Working set: " + p.WorkingSet64.ToString("#,#;;0"));
            Console.WriteLine("Managed mem: " + GC.GetTotalMemory(false).ToString("#,#;;0"));
            
        }
    }
}
```

**Comment by aviviadi** [comment]
What is worrying from our perspective is that we are policing ourselves in this case and reducing our total memory footprint, but we are unable to cause the system to release the memory from the working set so the OS won't consider us a good candidate for OOM execution.
